File: dart/Q31_submatrix_count/bin/solution1.dart
class Solution {
  int numSubmatrixSumTarget(List<List<int>> matrix, int target) {
    int count = 0;
    int rows = matrix.length;
    int cols = matrix[0].length;

    for (int top = 0; top < rows; top++) {
      for (int bottom = top; bottom < rows; bottom++) {
        for (int left = 0; left < cols; left++) {
          for (int right = left; right < cols; right++) {
            if (target == sum(matrix, top, bottom, left, right)) {
              count++;
            }
          }
        }
      }
    }
    return count;
  }

  int sum(List<List<int>> matrix, int top, int bottom, int left, int right) {
    int sum = 0;
    for (int r = top; r <= bottom; r++) {
      for (int c = left; c <= right; c++) {
        sum += matrix[r][c];
      }
    }
    return sum;
  }
}

File: ruby/Q30_subarray_sum/solution1.rb
def min_sub_array_len(target, nums)
  min_length = nums.length() + 1

  for left in 0..nums.length() - 1
    for right in left..nums.length() - 1
      sum = 0
      for i in left..right
        sum += nums[i]
      end
      if sum >= target && min_length > right - left + 1
        min_length = right - left + 1
        break
      end
    end
  end

  return min_length == nums.length() + 1 ? 0 : min_length
end
